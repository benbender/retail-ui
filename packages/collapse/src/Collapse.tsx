import { useThemeCtx } from '@retail-ui/theme'
import clsx from 'clsx'
import { constant } from 'lodash'
import * as React from 'react'

import { CollapseProvider } from './CollapseContext'

type ReactDivProps = React.HTMLAttributes<HTMLDivElement>
type Ref = HTMLDivElement

export type CollapseProps = {
  hasBorder?: boolean
}

export const Collapse = React.forwardRef<Ref, ReactDivProps & CollapseProps>(
  (props, ref) => {
    const { children, hasBorder, className, ...rest } = props

    const {
      theme: { CollapseStyles },
    } = useThemeCtx()

    const [collapses, setCollapses] = React.useState([
      ...[...Array(React.Children.count(children))].map(constant(false)),
    ])

    const renderChildren = React.useMemo(() => {
      return React.Children.map(children, (el, idx) => {
        return React.isValidElement(el)
          ? React.cloneElement(el, { index: `${idx}` })
          : null
      })
    }, [children])

    const callapseValue = React.useMemo(() => {
      return {
        collapses,
        setCollapses,
      }
    }, [collapses])

    const collapseCls = clsx(
      className,
      CollapseStyles.base,
      hasBorder && CollapseStyles.bordered,
    )

    return (
      <CollapseProvider value={callapseValue} {...rest}>
        <div ref={ref} className={collapseCls}>
          {renderChildren}
        </div>
      </CollapseProvider>
    )
  },
)

Collapse.displayName = 'Collapse'
