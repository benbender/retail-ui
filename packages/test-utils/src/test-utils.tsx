import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
