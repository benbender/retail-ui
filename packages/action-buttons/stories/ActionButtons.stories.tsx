import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
