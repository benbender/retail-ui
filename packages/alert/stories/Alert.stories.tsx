import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
