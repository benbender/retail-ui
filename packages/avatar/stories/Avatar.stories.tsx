import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
