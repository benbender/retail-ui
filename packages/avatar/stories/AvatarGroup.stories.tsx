import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
