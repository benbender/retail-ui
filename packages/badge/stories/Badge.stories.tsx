import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
