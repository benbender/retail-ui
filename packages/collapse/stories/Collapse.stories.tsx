import { Button } from '@retail-ui/button'
import * as React from 'react'

import { Collapse, CollapsePanel, CollapseProps } from '../src'

const meta = {
  title: 'Retail-UI/Collapse',
  component: Collapse,
}

export default meta

const Template = (args: CollapseProps) => (
  <Collapse {...args} data-testid={meta.title}>
    <CollapsePanel
      title="Open Collapse"
      buttonNode={({ onClick, isCollapsed }) => (
        <Button onClick={onClick}>
          {isCollapsed ? `Open collapse` : `Close collapse`}
        </Button>
      )}
    >
      Lorem, ipsum dolor sit amet consectetur adipisicing elit. Ipsum quaerat
      officia, atque iste ut excepturi minus? Tempora unde culpa odit
      dignissimos
    </CollapsePanel>
    <CollapsePanel title="Open Collapse">
      Lorem, ipsum dolor sit amet consectetur adipisicing elit. Ipsum quaerat
      officia, atque iste ut excepturi minus? Tempora unde culpa odit
      dignissimos
    </CollapsePanel>
    <CollapsePanel title="Open Collapse">
      Lorem, ipsum dolor sit amet consectetur adipisicing elit. Ipsum quaerat
      officia, atque iste ut excepturi minus? Tempora unde culpa odit
      dignissimos
    </CollapsePanel>
  </Collapse>
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
