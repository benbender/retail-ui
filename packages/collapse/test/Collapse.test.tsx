import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Collapse from '../stories/Collapse.stories'

describe('Collapse', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Collapse.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
