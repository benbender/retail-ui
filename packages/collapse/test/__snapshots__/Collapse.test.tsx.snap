// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders correctly 1`] = `
<DocumentFragment>
  <div
    class="divide-y divide-gray-200"
  >
    <div
      class="px-2 pb-3"
    >
      <button
        class="inline-flex items-center justify-center leading-5 align-bottom transition-colors duration-150 ease-in transform scale-100 border cursor-pointer focus:outline-none dark:focus:shadow-outline-gray active:scale-95 overflow-hidden no-underline select-none text-white bg-purple-600 border-transparent active:shadow-outline-purple focus:bg-purple-700 hover:bg-purple-700 py-2 px-4 text-base font-medium rounded-lg"
      >
        <span
          class="opacity-100"
        >
          Close collapse
        </span>
      </button>
    </div>
    <div
      class="px-2 pb-3"
    >
      <div
        aria-controls="CollapsePanel"
        aria-expanded="false"
        class="flex items-center justify-between p-3"
        role="button"
        tabindex="0"
      >
        Open Collapse
        <svg
          aria-hidden="true"
          class="w-7 h-7 text-current stroke-current transition ease-out duration-200 transform !h-5 !w-5 flex-none"
          fill="none"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
    <div
      class="px-2 pb-3"
    >
      <div
        aria-controls="CollapsePanel"
        aria-expanded="false"
        class="flex items-center justify-between p-3"
        role="button"
        tabindex="0"
      >
        Open Collapse
        <svg
          aria-hidden="true"
          class="w-7 h-7 text-current stroke-current transition ease-out duration-200 transform !h-5 !w-5 flex-none"
          fill="none"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
